Implemented fuzzy search #135

monarxxjain opened the pull request:
### Changes Description
Revamped icon search functionality with an advanced fuzzy search algorithm that vastly improves search
 accuracy and delivers better results. This optimized solution replaces the traditional search method, which
 couldn't handle misspellings or partial matches effectively. The new algorithm utilizes Monge-Elkan Algorithm, allowing users to find
 the exact icon they need quickly and effortlessly.

### NOTES
The match algorithm uses default value of THRESHOLD to be 0.8, which is fairly optimal to keep unnecessary clutter at bay, while maintaining context. with search text. If you want to be more pedantic with the search, then try increasing the threshold value to around 0.85 to 0.90.

### REFERENCES
To know more about Monge-Elkan Algorithm, refer linked references below:
https://link.springer.com/chapter/10.1007/978-3-642-00382-0_45
https://anhaidgroup.github.io/py_stringmatching/v0.3.x/MongeElkan.html
https://cs.stackexchange.com/questions/32530/a-reference-for-pseudocode-for-monge-elkan-algorithm

### Screenshots
<details >
<summary> Final Output </summary>

![peoplee](https://github.com/Bookingjini-Labs/bookingjini-icons/assets/128830789/a4e128af-6f39-47f0-9ab2-0eb0aeeda3cc)


![boot](https://github.com/Bookingjini-Labs/bookingjini-icons/assets/128830789/6c8eedf5-5cc4-4c91-b449-f122ee55c874)

![coin](https://github.com/Bookingjini-Labs/bookingjini-icons/assets/128830789/a256c35c-649e-406c-81d1-f1c28a0752fd)
</details>

<details>
<summary>Initial Output</summary>

![Current](https://github.com/Bookingjini-Labs/bookingjini-icons/assets/128830789/c93fe000-c944-46f1-bcbf-6e09345f99c7)

</details>






Fixes #135

monarxxjain committed:
Implemented fuzzy search

Revamped icon search functionality with an advanced fuzzy search algorithm that vastly improves search
 accuracy and delivers better results. This optimized solution replaces the traditional search method, which
 couldn't handle misspellings or partial matches effectively. The new algorithm utilizes Monge-Elkan Algorithm, allowing users to find
 the exact icon they need quickly and effortlessly.

AshutoshDash1999 commented:
@monarxxjain i am very much impressed with your PR. Can you take a pull and recheck if it's still working fine?

monarxxjain committed:
Merge branch 'Bookingjini-Labs:dev' into fuzzy

monarxxjain commented:
Fuzzy search implementation seems to work fine. Apart from some UI bugs in navbar and hero section, everything seems to work fine! However, those bugs are in the upstream repo itself, so nothing to be concerned of. 

However, one minor bug introduced due to #173,  **No result found** h4 tag is shown, even even element already exists.
https://github.com/Bookingjini-Labs/bookingjini-icons/blame/63735c0894b64fb1cd9890c233c51f2ed48ba0d9/website/index.html#L140


<details>
<summary>Screenshot</summary>
<img src="https://github.com/Bookingjini-Labs/bookingjini-icons/assets/128830789/4b7fa413-07b4-4e96-9432-ebb0b606e7a4" alt="Homepage">
</details>

monarxxjain committed:
Fixes bugs arised in #173

"No result found" h4 tag is shown, even even element already exists.
This commit fixes this. That would had worked fine with previous search algorithm, so I modified it to accommodate it  with current algo.

https://github.com/Bookingjini-Labs/bookingjini-icons/blame/63735c0894b64fb1cd9890c233c51f2ed48ba0d9/website/index.html#L140

monarxxjain commented:
Fixed that in last commit :-)
